examples/nemo/scripts/kernels_trans.py

```diff
@@ -226,98 +166,50 @@ def valid_acc_kernel(node):
                 continue
 
-            # Exclude things of the form IF(var == 0.0) because that causes
-            # deadlock in the code generated by the PGI compiler (<=19.4).
-            if PGI_VERSION <= 1940:
-                opn = enode.children[0]
-                if (excluding.ifs_real_scalars and
-                        isinstance(opn, BinaryOperation) and
-                        opn.operator == BinaryOperation.Operator.EQ and
-                        isinstance(opn.children[1], Literal) and
-                        opn.children[1].datatype.intrinsic ==
-                        ScalarType.Intrinsic.REAL):
-                    log_msg(routine_name,
-                            "IF performs comparison with REAL scalar", enode)
-                    return False
-
-            # We also permit single-statement IF blocks that contain a Loop
-            if "was_single_stmt" in enode.annotations and enode.walk(Loop):
-                continue
-            # When using CUDA managed memory, only allocated arrays are
-            # automatically put onto the GPU (although
-            # this includes those that are created by compiler-generated allocs
-            # e.g. for automatic arrays). We assume that all arrays of rank 2
-            # or greater are dynamically allocated.
-            arrays = enode.children[0].walk(ArrayReference)
-            # We also exclude if statements where the condition expression does
-            # not refer to arrays at all as this seems to cause issues for
-            # 19.4 of the compiler (get "Missing branch target block").
-            if not arrays and \
-               (excluding.ifs_scalars and not isinstance(enode.children[0],
-                                                         BinaryOperation)):
+            arrays = enode.condition.walk(ArrayReference)
+            # We exclude if statements where the condition expression does
+            # not refer to arrays at all as this may cause compiler issues
+            # (get "Missing branch target block") or produce faster code.
+            if not arrays and excluding.ifs_scalars and \
+               not isinstance(enode.condition, BinaryOperation):
                 log_msg(routine_name, "IF references scalars", enode)
                 return False
-            if excluding.ifs_1d_arrays and \
-               any([len(array.children) == 1 for array in arrays]):
+            # When using CUDA Unified Memory, only allocated arrays reside in
+            # shared memory (including those that are created by compiler-
+            # -generated allocs, e.g. for automatic arrays). We assume that all
+            # arrays of rank 2 or greater are dynamically allocated, whereas 1D
+            # arrays are often static in NEMO. Hence, we disallow IFs where the
+            # logical expression involves the latter.
+            if any([len(array.children) == 1 for array in arrays]):
```

Comment:
We could do better than this now for symbols defined in the current Fortran module. However, that can be for a subsequent PR.

Reply:
Also, pylint suggests using a generator here (which means just removing the `[]` in my experience).

Reply:
I'm guessing we could instead probe the static/dynamic status of arrays by checking for some keywords on the symbol table?

Reply:
Yes although it's slightly clunky now that I come to look at it - you have to look at the `shape` of the datatype of the Symbol.

Reply:
Ah... so I'd still be reasoning on the number of ranks of the involved arrays, not on keywords, like `allocatable` or `pointer`...